### PR TITLE
IDENTITY-3713 Same username(user1) can be added before delete the user(user1) via workflow

### PR DESCRIPTION
**Pull request by cdwijayarathna**

### CODE REVIEW

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java [hunk 1]
```diff
@@ -214,19 +216,42 @@ public void onWorkflowCompletion(String status, Map<String, Object> requestParam
...
                         .entityHasPendingWorkflowsOfType(entities[i], UserStoreWFConstants.ADD_ROLE_EVENT) ||
                         workflowService.entityHasPendingWorkflowsOfType(entities[i], UserStoreWFConstants
-                                .UPDATE_ROLE_NAME_EVENT))) {
-                    throw new WorkflowException("Rolename already exists in the system. Please pick another rolename.");
+                                .UPDATE_ROLE_NAME_EVENT) || userStoreManager.isExistingRole(entities[i].getEntityId()
+                ))) {
+
+                    throw new WorkflowException("Role name already exists in the system. Please pick another role " +
+                            "name.");
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java [hunk 2]
```diff
@@ -214,19 +216,42 @@ public void onWorkflowCompletion(String status, Map<String, Object> requestParam
...
         for (int i = 0; i < entities.length; i++) {
             try {
                 if (entities[i].getEntityType() == UserStoreWFConstants.ENTITY_TYPE_ROLE && (workflowService
                         .entityHasPendingWorkflowsOfType(entities[i], UserStoreWFConstants.ADD_ROLE_EVENT) ||
                         workflowService.entityHasPendingWorkflowsOfType(entities[i], UserStoreWFConstants
-                                .UPDATE_ROLE_NAME_EVENT))) {
-                    throw new WorkflowException("Rolename already exists in the system. Please pick another rolename.");
+                                .UPDATE_ROLE_NAME_EVENT) || userStoreManager.isExistingRole(entities[i].getEntityId()
+                ))) {
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java [hunk 3]
```diff
@@ -214,19 +216,42 @@ public void onWorkflowCompletion(String status, Map<String, Object> requestParam
...
+                    throw new WorkflowException("Role name already exists in the system. Please pick another role " +
+                            "name.");
+
                 } else if (workflowService.eventEngagedWithWorkflows(UserStoreWFConstants.ADD_USER_EVENT) &&
                         entities[i].getEntityType() == UserStoreWFConstants.ENTITY_TYPE_USER && workflowService
                         .entityHasPendingWorkflowsOfType(entities[i], UserStoreWFConstants.DELETE_USER_EVENT)) {
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java [hunk 4]
```diff
@@ -214,19 +216,42 @@ public void onWorkflowCompletion(String status, Map<String, Object> requestParam
...
                     throw new WorkflowException("One or more assigned users are pending in delete workflow.");
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java [hunk 5]
```diff
@@ -214,19 +216,42 @@ public void onWorkflowCompletion(String status, Map<String, Object> requestParam
...
+                } else if (entities[i].getEntityType() == UserStoreWFConstants.ENTITY_TYPE_USER && !userStoreManager
+                        .isExistingUser(entities[i].getEntityId())) {
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java [hunk 6]
```diff
@@ -214,19 +216,42 @@ public void onWorkflowCompletion(String status, Map<String, Object> requestParam
...
+                    throw new WorkflowException("User " + entities[i].getEntityId() + " does not exist.");
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java [hunk 7]
```diff
@@ -252,12 +254,26 @@ public void onWorkflowCompletion(String status, Map<String, Object> requestParam
...
+            throw new WorkflowException("Error while retrieving user realm.", e);
+        }
         for (int i = 0; i < entities.length; i++) {
             try {
-                if (entities[i].getEntityType() == UserStoreWFConstants.ENTITY_TYPE_USER && workflowService
-                        .entityHasPendingWorkflowsOfType(entities[i], UserStoreWFConstants.ADD_USER_EVENT)) {
+                if (entities[i].getEntityType() == UserStoreWFConstants.ENTITY_TYPE_USER && (workflowService
+                        .entityHasPendingWorkflowsOfType(entities[i], UserStoreWFConstants.ADD_USER_EVENT) ||
+                        userStoreManager.isExistingUser(entities[i].getEntityId()))) {
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java [hunk 8]
```diff
@@ -266,8 +282,15 @@ public boolean isValidOperation(Entity[] entities) throws WorkflowException {
                     throw new WorkflowException("One or more roles assigned has pending workflows which " +
                             "blocks this operation.");
 
+                } else if (eventEngaged && (entities[i].getEntityType() == UserStoreWFConstants
+                        .ENTITY_TYPE_ROLE && !userStoreManager.isExistingRole(entities[i].getEntityId()))) {
+
+                    //Check condition only when event engaged to workflow since failure at populating users for tests.
+                    throw new WorkflowException("Role " + entities[i].getEntityId() + " does not exist.");
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java [hunk 9]
```diff
@@ -266,8 +282,15 @@ public boolean isValidOperation(Entity[] entities) throws WorkflowException {
...
                 }
-            } catch (InternalWorkflowException e) {
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteClaimWFRequestHandler.java [hunk 10]
```diff
@@ -186,6 +199,10 @@ public boolean isValidOperation(Entity[] entities) throws WorkflowException {
 
                     throw new WorkflowException("User has a delete operation pending.");
 
+                } else if (entities[i].getEntityType() == UserStoreWFConstants.ENTITY_TYPE_USER && !userStoreManager
+                        .isExistingUser(entities[i].getEntityId())) {
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteClaimWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteMultipleClaimsWFRequestHandler.java [hunk 11]
```diff
@@ -183,12 +185,27 @@ public String getCategory() {
...
+
         for (int i = 0; i < entities.length; i++) {
             try {
                 if (entities[i].getEntityType() == UserStoreWFConstants.ENTITY_TYPE_USER && workflowService
                         .entityHasPendingWorkflowsOfType(entities[i], UserStoreWFConstants.DELETE_USER_EVENT)) {
 
                     throw new WorkflowException("User has a delete operation pending.");
+                } else if (entities[i].getEntityType() == UserStoreWFConstants.ENTITY_TYPE_USER && !userStoreManager
+                        .isExistingUser(entities[i].getEntityId())) {
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteMultipleClaimsWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/SetMultipleClaimsWFRequestHandler.java [hunk 12]
```diff
@@ -181,12 +183,27 @@ public String getCategory() {
...
+
         for (int i = 0; i < entities.length; i++) {
             try {
                 if (entities[i].getEntityType() == UserStoreWFConstants.ENTITY_TYPE_USER && workflowService
                         .entityHasPendingWorkflowsOfType(entities[i], UserStoreWFConstants.DELETE_USER_EVENT)) {
 
                     throw new WorkflowException("User has a delete operation pending.");
+                } else if (entities[i].getEntityType() == UserStoreWFConstants.ENTITY_TYPE_USER && !userStoreManager
+                        .isExistingUser(entities[i].getEntityId())) {
+
```
Comment (pulasthi7) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/SetMultipleClaimsWFRequestHandler.java]: unnecessary new line

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java (hunk 1 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java (hunk 2 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java (hunk 3 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java (hunk 4 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java (hunk 5 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java (hunk 6 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddRoleWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java (hunk 7 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java (hunk 8 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java (hunk 9 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/AddUserWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteClaimWFRequestHandler.java (hunk 10 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteClaimWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteMultipleClaimsWFRequestHandler.java (hunk 11 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/DeleteMultipleClaimsWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/SetMultipleClaimsWFRequestHandler.java (hunk 12 above)
Comment (cdwijayarathna) [inline components/user-mgt/org.wso2.carbon.user.mgt.workflow/src/main/java/org/wso2/carbon/user/mgt/workflow/userstore/SetMultipleClaimsWFRequestHandler.java]: Fixed with https://github.com/wso2/carbon-identity/pull/1675

**Pull request merged**